fvm/evm/precompiles/abi.go

```diff
@@ -0,0 +1,223 @@
+	if len(buffer) < index+Bytes4DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	// fixed-size byte values are zero-padded on the right side.
+	return buffer[index : index+Bytes4DataReadSize], nil
+}
+
+// ReadBytes8 reads a 8 byte slice from the buffer at index
+func ReadBytes8(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+Bytes8DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	// fixed-size byte values are zero-padded on the right side.
+	return buffer[index : index+Bytes8DataReadSize], nil
+}
+
+// ReadBytes32 reads a 32 byte slice from the buffer at index
+func ReadBytes32(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+Bytes32DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	return buffer[index : index+Bytes32DataReadSize], nil
+}
+
+// EncodeBytes32 encodes data into a bytes 32
+func EncodeBytes32(data []byte, buffer []byte, index int) error {
+	if len(data) > EncodedBytes32Size {
+		return ErrDataTooLarge
+	}
+	if len(buffer) < index+EncodedBytes32Size {
+		return ErrBufferTooSmall
+	}
+	copy(buffer[index:index+EncodedBytes32Size],
+		gethCommon.RightPadBytes(data, EncodedBytes32Size),
+	)
+	return nil
+}
+
+// ReadBytes reads a variable length bytes from the buffer
+func ReadBytes(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+EncodedUint64Size {
+		return nil, ErrInputDataTooSmall
+	}
+	// reading offset (we read into uint64) and adjust index
+	offset, err := ReadUint64(buffer, index)
+	if err != nil {
+		return nil, err
+	}
+	index = int(offset)
+	if len(buffer) < index+EncodedUint64Size {
+		return nil, ErrInputDataTooSmall
+	}
+	// reading length of byte slice
+	length, err := ReadUint64(buffer, index)
+	if err != nil {
+		return nil, err
+	}
+	index += EncodedUint64Size
+	if len(buffer) < index+int(length) {
+		return nil, ErrInputDataTooSmall
+	}
+	return buffer[index : index+int(length)], nil
+}
+
+// SizeNeededForBytesEncoding computes the number of bytes needed for bytes encoding
+func SizeNeededForBytesEncoding(data []byte) int {
+	paddedSize := (len(data) / FixedSizeUnitDataReadSize) + FixedSizeUnitDataReadSize
+	return EncodedUint64Size + EncodedUint64Size + paddedSize
+}
+
+// EncodeBytes encodes the data into the buffer at index and append payload to the
+// end of buffer
+func EncodeBytes(data []byte, buffer []byte, headerIndex, payloadIndex int) error {
+	//// updating offset
+	if len(buffer) < headerIndex+EncodedUint64Size {
+		return ErrBufferTooSmall
+	}
+	err := EncodeUint64(uint64(payloadIndex), buffer, headerIndex)
+	if err != nil {
+		return err
+	}
+	headerIndex += EncodedUint64Size
+
+	//// updating payload
+	// padding data
+	dataSize := len(data)
+	paddedSize := (dataSize / FixedSizeUnitDataReadSize) + FixedSizeUnitDataReadSize
+	if dataSize%FixedSizeUnitDataReadSize != 0 {
+		data = gethCommon.RightPadBytes(data, paddedSize)
+	}
+	if len(buffer) < payloadIndex+EncodedUint64Size+paddedSize {
```

Comment:
wouldn't it be better for this if statement to be moved at least on line 208 if not even before on 197 to avoid any changes to the data if the buffer is too small?

Reply:
moved it